### DISCUSSION TITLE
Why interface conditions degrade convergence by an order?

### DISCUSSION THREAD

**Discussion by lindsayad** [idaholab/moose#22808] [Q&A General] [answered]
When working on https://github.com/idaholab/moose/pull/22130 I am presented with a classic conundrum. How to enforce the Bernoulli condition at a porosity discontinuity (interface)? There are a few ways:

1. Lagrange multiplier
2. penalty
3. "Matched" value (where the value is computed from Bernoulli's equation) and matched flux condition

The third option is very nice in that it doesn't introduce additional degrees of freedom and degrees of freedom which may not have diagonals (the LM approach) nor does it introduce ill-conditioning through a penalty term. However, [in a thread](https://groups.google.com/g/moose-users/c/fi68gQRZl9g/m/nwSv-EDfBgAJ) from four years ago @YaqiWang helped me discover that the third option introduces an order reduction in the convergence rate. My question is do we know a fundamental reason why this reduction occurs? Is it the asymmetry of the approach? Looking for help from mathy people. Maybe @roystgnr @hugary1995 @oanaoana ?

**Comment by oanaoana** [comment]
Do you have a proper description of the problem. Which equation is solved and how? Or what code exactly implements this? to track it down that way. Looking through your links to understand it better.

**Comment by roystgnr** [comment, marked as answer]
Completely wild guess: could you effectively be getting rid of some degrees of freedom by trying to match both values and fluxes too strongly?  I'm not a FVM guy at all, but at first glance this idea vaguely reminds me of the typical intro-FEM mistake, where you pin down a Neumann BC exactly instead of imposing it weakly and then it turns out that the price for getting the flux exactly right on that boundary is getting an inferior solution over the whole domain.

Or I could be completely off-base.  In the strong-Neumann-FEM setting it's easy to see that the error comes from thinking of the approximate flux as being equally correct everywhere (or at least on the element boundary), instead of only being trustworthy at superconvergent points or after some kind of recovery procedure is used.  When I think about it that way I don't see how you could be making the analogous mistake here.  I'll second Oana's request.

**Reply by lindsayad** [reply]
In the four year old thread, the problem is a finite element problem. The finite element problem was this essentially: solve diffusion (Laplacian) on left and right domains for variables u and v. At a conforming interface in the middle, u and v should be equal and the diffusive fluxes should be equal. At the left boundary u is 1; at the right boundary v is 0.

**Reply by lindsayad** [reply]
In the implementation that showed the order reduction in convergence, v was strongly set to u at the interface. The diffusive flux of u was weakly set to the diffusive flux of v at the interface.

**Reply by roystgnr** [reply]
Those plots don't look like harmonic functions.  Is this Poisson, or transient Laplacian?

How are you setting the diffusive flux?  Evaluating k∇v on one side and using it as a weak Neumann condition for u, while also using the value of u on the other side as a strong Dirichlet condition for v?

**Reply by lindsayad** [reply]
> Those plots don't look like harmonic functions. Is this Poisson, or transient Laplacian?

Oops yea since I was solving an MMS problem, it's Poisson.

> How are you setting the diffusive flux? Evaluating k∇v on one side and using it as a weak Neumann condition for u, while also using the value of u on the other side as a strong Dirichlet condition for v?

That's exactly correct. How did you do that nabla symbol BTW?

**Reply by roystgnr** [reply]
> That's exactly correct.

That would be enough to kill your order, I think.  Imagine solving Δu=f, Δv=f that way, on a line segment split half-and-half, with constant f.  I'd have to work it out to say exactly how it would shake out, but at the interface if ∇v=0 then you're not at the optimal solution for v (which should have a slight gradient on the interface element because it's the *post-processed* flux at the vertex that should be 0, not the one-sided flux from either element) and if ∇v≠0 then you're not at the optimal solution for u (because you would get that by imposing 0 flux weakly).

> How did you do that nabla symbol BTW?

Windows key, m, D.  ("math, Delta" mnemonic; Win,g,D would have given Δ for "greek, Delta").  I tricked out my .XCompose file a while back.  I never managed to get good at taking notes with LaTeX notation, but *unicode* math notation is pretty sweet 95% of the time.  I even went back through old notes in vim, to search and replace stuff like R^n → ℝⁿ.  This seems to be a little hit or miss on web fonts, though.  I dunno about your browser but my copy of Chrome is not rendering https://www.compart.com/en/unicode/U+2260 right in either the Write or Preview window...

**Reply by GiudGiud** [reply]
the computation of flux/current usually involves the boundary value, so imposing one may impose the other to a certain order. I dont think that is a general rule, it's true of some discretization schemes. 

For example for a basic 1D finite difference diffusion, the strong continuity of current will give you:
(E F C for elem1 centroid, face, elem2 centroid, a is the variable, the current is the 1st order one sided da/dx times a diffusion coefficient.)
D1 (af - ae) / dx = D2 (ac - af) / dx
or af = (ae D1 + ac D2) / (D1 + D2)
so that imposed the face value.

But this doesnt really prove my point even if we expand the continuity of value equation because the scheme seems 1st order to start with. I seem to get a different imposed face value from the continuity of value equation, with a first order error, so far

For finite volume, if this is coming up, we definitely want to prioritize continuity of current/flux over continuity of the values. We care more about preserving the balance of the actual physical quantity than its values at any location imo.
Now one thing that could be tricky would be: if we have to choose between conserving fluxes of mass, momentum and/or advected quantities at an interface. I m speculating at this point

**Reply by hugary1995** [reply]
For enforcing general nonlinear jump condition across an interface, I prefer Nitsche > penalty ~= LM. You are apparently aware of the drawbacks of each approach. The matched-value approach (mixed Dirichlet and Neumann across the interface) will indeed give you a suboptimal order of convergence.

**Reply by jiangwen84** [reply]
I think the key to get the optimal rate of convergence is the approach needs to be variational consistent.  For the matched value approach,  the Dirichlet part does not seem to be variational consistent.  For that perspective,  I agree with Gary that Nitshce is most preferred.

**Reply by lindsayad** [reply]
Thanks to all of you for your weigh-in. Weak enforcement definitely seems like the way to go! Luckily for me, the current PR that is motivating me is with finite volume in which we always impose Dirichlet conditions weakly. Per my most recent commit https://github.com/idaholab/moose/pull/22130/commits/9b516099dfc27ed3a1d0e1b0a7324b5bacdce2bd I'm able to maintain the optimal second order convergence rate using a "naive" matching value/matching flux approach.

**Reply by lindsayad** [reply]
> For finite volume, if this is coming up, we definitely want to prioritize continuity of current/flux over continuity of the values

In this discussion I regard enforcing continuity as a conceptual surrogate for what I will want to do in #22130 which is enforcing Bernoulli, so matching values with an offset

**Reply by JerryLiu2023** [reply]
matching "flux" and matching 'primary solution' could be very different? flux computed already lost one order?

**Reply by lindsayad** [reply]
The lost order is with respect to an L2 error measurement against an MMS analytic solution
